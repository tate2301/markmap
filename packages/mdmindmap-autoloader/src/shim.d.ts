declare interface Window {
  markmap: typeof import('mdmindmap-lib') &
    typeof import('mdmindmap-view') &
    typeof import('mdmindmap-toolbar') & {
      autoLoader?: Partial<import('.').AutoLoaderOptions>;
    };
}
