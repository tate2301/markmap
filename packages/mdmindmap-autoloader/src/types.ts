import { JSItem, CSSItem } from 'mdmindmap-common';
import { ITransformPlugin } from 'mdmindmap-lib';

export interface AutoLoaderOptions {
  baseJs: (string | JSItem)[];
  baseCss: (string | CSSItem)[];
  provider: string | ((path: string) => string);
  /** Callback when mdmindmap-lib/mdmindmap-view and their dependencies are loaded. We can tweak global options in this callback. */
  onReady: () => void;
  /** Override built-in plugins if provided. Set to `[]` to disable all built-in plugins for auto-loader. */
  transformPlugins: Array<ITransformPlugin | (() => ITransformPlugin)>;
  /** Whether to render markmaps manually. If false, all elements matching `.markmap` will be rendered once this package loads or DOMContentLoaded is emitted, whichever later. */
  manual: boolean;
  /** Whether to create a toolbar for each markmap. */
  toolbar: boolean;
}
