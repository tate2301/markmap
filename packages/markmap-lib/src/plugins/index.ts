import pluginCheckbox from './checkbox';
import pluginFrontmatter from './frontmatter';
import pluginHljs from './hljs';
import pluginKatex from './katex';
import pluginNpmUrl from './npm-url';
import pluginSourceLines from './source-lines';
import pluginBidirectional from './bidirectional';

export * from './base';

export {
  pluginCheckbox,
  pluginFrontmatter,
  pluginHljs,
  pluginKatex,
  pluginNpmUrl,
  pluginSourceLines,
  pluginBidirectional,
};

export const plugins = [
  pluginFrontmatter,
  pluginKatex,
  pluginHljs,
  pluginNpmUrl,
  pluginCheckbox,
  pluginSourceLines,
  pluginBidirectional,
];
