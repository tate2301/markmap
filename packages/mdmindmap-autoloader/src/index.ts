import {
  buildCSSItem,
  buildJSItem,
  loadCSS,
  loadJS,
  urlBuilder,
} from 'mdmindmap-common';
import type { Transformer } from 'mdmindmap-lib';
import type { AutoLoaderOptions } from './types';

export * from './types';

const enabled: Record<string, boolean> = {};

const autoLoaderOptions = {
  baseJs: [
    `d3@${__define__.D3_VERSION}`,
    `mdmindmap-lib@${__define__.LIB_VERSION}`,
    `mdmindmap-view@${__define__.VIEW_VERSION}`,
    `mdmindmap-toolbar@${__define__.TOOLBAR_VERSION}`,
  ],
  baseCss: [`mdmindmap-toolbar@${__define__.TOOLBAR_VERSION}/dist/style.css`],
  manual: false,
  toolbar: false,
  ...(window.markmap?.autoLoader as Partial<AutoLoaderOptions>),
};

async function initialize() {
  if (typeof autoLoaderOptions.provider === 'function') {
    urlBuilder.setProvider(
      (urlBuilder.provider = 'autoLoader'),
      autoLoaderOptions.provider,
    );
  } else if (typeof autoLoaderOptions.provider === 'string') {
    urlBuilder.provider = autoLoaderOptions.provider;
  } else {
    try {
      await urlBuilder.findFastestProvider();
    } catch {
      // ignore
    }
  }
  await Promise.all([
    loadJS(
      autoLoaderOptions.baseJs.map((item) =>
        typeof item === 'string'
          ? buildJSItem(urlBuilder.getFullUrl(item))
          : item,
      ),
    ),
    loadCSS(
      autoLoaderOptions.baseCss.map((item) =>
        typeof item === 'string'
          ? buildCSSItem(urlBuilder.getFullUrl(item))
          : item,
      ),
    ),
  ]);
  const { markmap } = window;
  const style = document.createElement('style');
  style.textContent = markmap.globalCSS;
  // Insert global CSS to body so it has higher priority than prism.css, etc.
  document.body.prepend(style);
  autoLoaderOptions.onReady?.();
}

export const ready = initialize();

function transform(transformer: Transformer, content: string) {
  const result = transformer.transform(content);
  const keys = Object.keys(result.features).filter((key) => !enabled[key]);
  keys.forEach((key) => {
    enabled[key] = true;
  });
  const { styles, scripts } = transformer.getAssets(keys);
  const { markmap } = window;
  if (styles) markmap.loadCSS(styles);
  if (scripts) markmap.loadJS(scripts);
  return result;
}

export function render(el: HTMLElement) {
  const { Transformer, Markmap, deriveOptions, Toolbar } = window.markmap;
  const lines = el.textContent?.split('\n') || [];
  let indent = Infinity;
  lines.forEach((line) => {
    const spaces = line.match(/^\s*/)?.[0].length || 0;
    if (spaces < line.length) indent = Math.min(indent, spaces);
  });
  const content = lines
    .map((line) => line.slice(indent))
    .join('\n')
    .trim();
  const transformer = new Transformer(autoLoaderOptions.transformPlugins);
  transformer.urlBuilder = urlBuilder;
  el.innerHTML = '<svg></svg>';
  const svg = el.firstChild as SVGElement;
  const mm = Markmap.create(svg, { embedGlobalCSS: false });
  if (autoLoaderOptions.toolbar) {
    const { el: toolbar } = Toolbar.create(mm);
    Object.assign(toolbar.style, {
      position: 'absolute',
      right: '20px',
      bottom: '20px',
    });
    el.append(toolbar);
  }
  const doRender = async () => {
    const { root, frontmatter } = transform(transformer, content);
    const markmapOptions = frontmatter?.markmap;
    const frontmatterOptions = deriveOptions(markmapOptions);
    await mm.setData(root, frontmatterOptions);
    mm.fit();
  };
  transformer.hooks.retransform.tap(doRender);
  doRender();
}

export async function renderAllUnder(container: ParentNode) {
  await ready;
  container.querySelectorAll<HTMLElement>('.markmap').forEach(render);
}

export function renderAll() {
  return renderAllUnder(document);
}

if (!autoLoaderOptions.manual) {
  if (document.readyState === 'loading')
    document.addEventListener('DOMContentLoaded', () => {
      renderAll();
    });
  else renderAll();
}
