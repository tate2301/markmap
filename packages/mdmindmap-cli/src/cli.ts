import { program } from 'commander';
import { readFile } from 'fs/promises';
import open from 'open';
import { basename } from 'path';
import { readPackageUp } from 'read-package-up';
import updateNotifier from 'update-notifier';
import { fileURLToPath } from 'url';
import { createMarkmap } from '.';
import { fetchAssets } from './util/assets-fetcher';
import { develop } from './util/dev-server';

async function main() {
  const pkg = (
    await readPackageUp({
      cwd: fileURLToPath(import.meta.url),
    })
  )?.packageJson;
  if (!pkg) throw new Error('Invalid package');

  const notifier = updateNotifier({ pkg });
  notifier.notify();

  program
    .version(pkg.version)
    .description('Create a markmap from a Markdown input file')
    .arguments('<input>')
    .option('--no-open', 'Do not open the output file after generation')
    .option('--no-toolbar', 'Do not show toolbar')
    .option('-o, --output <output>', 'Specify the filename of the output HTML')
    .option(
      '--offline',
      'Inline all assets to allow the generated HTML to work offline',
    )
    .option(
      '-w, --watch',
      'Watch the input file and update output on the fly, note that this feature is for development only',
    )
    .option('--port <port>', 'Set the port for the devServer to listen')
    .action(async (input: string, cmd) => {
      let { offline } = cmd;
      if (cmd.watch) offline = true;
      if (offline) await fetchAssets();
      const content = await readFile(input, 'utf8');
      const output = cmd.output || `${input.replace(/\.\w*$/, '')}.html`;
      if (cmd.watch) {
        const devServer = await develop({
          toolbar: cmd.toolbar,
          offline,
          port: +cmd.port || undefined,
        });
        const address = devServer.serverInfo!.address;
        const provider = devServer.addProvider({ filePath: input });
        const filename = basename(input);
        const url = `http://localhost:${address.port}/?key=${provider.key}&filename=${encodeURIComponent(filename)}`;
        console.log(`Listening at ${url}`);
        if (cmd.open) open(url);
      } else {
        await createMarkmap({
          content,
          output,
          open: cmd.open,
          toolbar: cmd.toolbar,
          offline,
        });
      }
    });
  program.parse();
}

main().catch((err) => {
  console.error(err);
  process.exitCode = 1;
});
