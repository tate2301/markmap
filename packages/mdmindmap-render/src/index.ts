import {
  IAssets,
  IPureNode,
  JSItem,
  UrlBuilder,
  buildJSItem,
  urlBuilder as defaultUrlBuilder,
  persistCSS,
  persistJS,
} from 'mdmindmap-common';
import type { IMarkmapJSONOptions, IMarkmapOptions } from 'mdmindmap-view';

export const template = __define__.TEMPLATE || '';

export const baseJsPaths = [
  `d3@${__define__.D3_VERSION}/dist/d3.min.js`,
  `mdmindmap-view@${__define__.VIEW_VERSION}/dist/browser/index.js`,
];

export function fillTemplate(
  root: IPureNode | null,
  assets: IAssets,
  extra?: {
    baseJs?: JSItem[];
    jsonOptions?: Partial<IMarkmapJSONOptions>;
    getOptions?: (
      jsonOptions: Partial<IMarkmapJSONOptions>,
    ) => Partial<IMarkmapOptions>;
    /** See https://github.com/gera2ld/npm2url */
    urlBuilder?: UrlBuilder;
  },
): string {
  extra = {
    ...extra,
  };
  const urlBuilder = extra.urlBuilder ?? defaultUrlBuilder;
  extra.baseJs ??= baseJsPaths
    .map((path) => urlBuilder.getFullUrl(path))
    .map((path) => buildJSItem(path));
  const { scripts, styles } = assets;
  const cssList = [...(styles ? persistCSS(styles) : [])];
  const context = {
    getMarkmap: () => window.markmap,
    getOptions: extra.getOptions,
    jsonOptions: extra.jsonOptions,
    root,
  };
  const jsList = [
    ...persistJS(
      [
        ...extra.baseJs,
        ...(scripts || []),
        {
          type: 'iife',
          data: {
            fn: (
              getMarkmap: (typeof context)['getMarkmap'],
              getOptions: (typeof context)['getOptions'],
              root: (typeof context)['root'],
              jsonOptions: IMarkmapJSONOptions,
            ) => {
              const markmap = getMarkmap();
              window.mm = markmap.Markmap.create(
                'svg#mindmap',
                (getOptions || markmap.deriveOptions)(jsonOptions),
                root,
              );
            },
            getParams: ({ getMarkmap, getOptions, root, jsonOptions }) => {
              return [getMarkmap, getOptions, root, jsonOptions];
            },
          },
        } as JSItem,
      ],
      context,
    ),
  ];
  const html = template
    .replace('<!--CSS-->', () => cssList.join(''))
    .replace('<!--JS-->', () => jsList.join(''));
  return html;
}
