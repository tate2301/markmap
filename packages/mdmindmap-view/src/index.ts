export { loadJS, loadCSS } from 'mdmindmap-common';

export * from './types';
export * from './constants';
