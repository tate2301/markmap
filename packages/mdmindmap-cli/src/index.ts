import { readFile, writeFile } from 'fs/promises';
import { CSSItem, JSItem, buildJSItem, mergeAssets } from 'mdmindmap-common';
import {
  Transformer,
  type IAssets,
  type IMarkmapCreateOptions,
} from 'mdmindmap-lib';
import { baseJsPaths, fillTemplate } from 'mdmindmap-render';
import open from 'open';
import { resolve } from 'path';
import { IDevelopOptions } from './types';
import {
  ASSETS_PREFIX,
  config,
  localProvider,
  toolbarAssets,
} from './util/common';

export * from './types';
export * from './util/dev-server';
export { config };

export * as markmap from 'mdmindmap-lib';

async function loadFile(path: string) {
  if (path.startsWith(ASSETS_PREFIX)) {
    const relpath = path.slice(ASSETS_PREFIX.length);
    return readFile(resolve(config.assetsDir, relpath), 'utf8');
  }
  const res = await fetch(path);
  if (!res.ok) throw res;
  return res.text();
}

async function inlineAssets(assets: IAssets): Promise<IAssets> {
  const [scripts, styles] = await Promise.all([
    Promise.all(
      (assets.scripts || []).map(
        async (item): Promise<JSItem> =>
          item.type === 'script' && item.data.src
            ? {
                type: 'script',
                data: {
                  textContent: await loadFile(item.data.src),
                },
              }
            : item,
      ),
    ),
    Promise.all(
      (assets.styles || []).map(
        async (item): Promise<CSSItem> =>
          item.type === 'stylesheet'
            ? {
                type: 'style',
                data: await loadFile(item.data.href),
              }
            : item,
      ),
    ),
  ]);
  return {
    scripts,
    styles,
  };
}

export async function createMarkmap(
  options: IMarkmapCreateOptions & IDevelopOptions & { open: boolean },
): Promise<void> {
  const transformer = new Transformer();
  if (options.offline) {
    transformer.urlBuilder.setProvider('local', localProvider);
    transformer.urlBuilder.provider = 'local';
  } else {
    try {
      await transformer.urlBuilder.findFastestProvider();
    } catch {
      // ignore
    }
  }
  const { root, features, frontmatter } = transformer.transform(
    options.content || '',
  );
  const otherAssets = mergeAssets(
    {
      scripts: baseJsPaths.map(buildJSItem),
    },
    options.toolbar ? toolbarAssets : null,
  );
  let assets = mergeAssets(
    {
      scripts: otherAssets.scripts?.map((item) => transformer.resolveJS(item)),
      styles: otherAssets.styles?.map((item) => transformer.resolveCSS(item)),
    },
    transformer.getUsedAssets(features),
  );
  if (options.offline) assets = await inlineAssets(assets);
  const html = fillTemplate(root, assets, {
    baseJs: [],
    jsonOptions: (frontmatter as any)?.markmap,
    urlBuilder: transformer.urlBuilder,
  });
  const output = options.output || 'markmap.html';
  await writeFile(output, html, 'utf8');
  if (options.open) open(output);
}
