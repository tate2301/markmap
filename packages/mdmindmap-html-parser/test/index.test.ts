import { expect, test } from 'vitest';
import { convertNode, parseHtml } from '../src/index';

test('parseHtml', () => {
  const root = parseHtml(`
<body>
<div class="container">
<h1 data-id="h1">heading 1</h1>
<p>text here will be ignored.</p>
<ul>
<li>this list</li>
<li>is ignored</li>
</ul>
<h2 data-id="h2">heading 2 <!-- markmap: foldAll --></h2>
<p>text also ignored</p>
<ul>
<li><p>item 1</p></li>
<li><p>item 2</p>
<ul>
<li><p>item 3</p></li>
<li><p>item 4</p><p>additional text</p></li>
</li>
</ul>
</ul>
<ol>
<li>item 5</li>
<li>item 6</li>
<li>item 7
<ul>
<li>item 7.1
<ul>
<li>item 7.1.1</li>
<li>item 7.1.2</li>
</ul>
</li>
<li>item 7.2</li>
</ul>
</li>
</ol>
<h2>A table</h2>
<table>
<tr>
<td>
<ul><li>this list is ignored</li></ul>
</td>
</tr>
</table>
<p><img src="image1.png"></p>
<p><img src="image2.png"></p>
</div>
</body>
`);
  expect(root).toMatchSnapshot();
  expect(convertNode(root)).toMatchSnapshot();
});

test('parseHtml with data', () => {
  const root = parseHtml(`
<ul data-lines="0,4">
<li data-lines="0,4">l1
<ul data-lines="1,4">
<li data-lines="1,2">l1.1</li>
<li data-lines="2,4">l1.2
<ul data-lines="3,4">
<li data-lines="3,4">l1.2.1</li>
</ul></li>
</ul></li>
</ul>`);
  expect(root).toMatchSnapshot();
  expect(convertNode(root)).toMatchSnapshot();
});

test('li > pre', () => {
  const root = parseHtml(`<body>
<ul>
<li>hello</li>
<li><pre><code>code block</code></pre></li>
</ul>
</body>`);
  expect(root).toMatchSnapshot();
  expect(convertNode(root)).toMatchSnapshot();
});

test('ol > li', () => {
  const root = parseHtml(`<body>
<ol start="3">
<li>hello</li>
<li>world</li>
</ol>
</body>`);
  expect(root).toMatchSnapshot();
  expect(convertNode(root)).toMatchSnapshot();
});
