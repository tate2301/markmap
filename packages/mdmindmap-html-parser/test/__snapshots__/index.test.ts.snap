// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`li > pre 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "childrenLevel": 0,
          "html": "hello",
          "id": 2,
          "level": 9,
          "parent": 1,
          "tag": "li",
        },
        {
          "children": [],
          "childrenLevel": 0,
          "html": "<pre><code>code block</code></pre>",
          "id": 3,
          "level": 9,
          "parent": 1,
          "tag": "li",
        },
      ],
      "childrenLevel": 9,
      "html": "",
      "id": 1,
      "level": 8,
      "parent": 0,
      "tag": "ul",
    },
  ],
  "childrenLevel": 8,
  "html": "",
  "id": 0,
  "level": 0,
  "parent": 0,
  "tag": "",
}
`;

exports[`li > pre 2`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "content": "hello",
        },
        {
          "children": [],
          "content": "<pre><code>code block</code></pre>",
        },
      ],
      "content": "",
    },
  ],
  "content": "",
}
`;

exports[`ol > li 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "childrenLevel": 0,
          "data": {
            "listIndex": 3,
          },
          "html": "3. hello",
          "id": 2,
          "level": 9,
          "parent": 1,
          "tag": "li",
        },
        {
          "children": [],
          "childrenLevel": 0,
          "data": {
            "listIndex": 4,
          },
          "html": "4. world",
          "id": 3,
          "level": 9,
          "parent": 1,
          "tag": "li",
        },
      ],
      "childrenLevel": 9,
      "html": "",
      "id": 1,
      "level": 8,
      "parent": 0,
      "tag": "ol",
    },
  ],
  "childrenLevel": 8,
  "html": "",
  "id": 0,
  "level": 0,
  "parent": 0,
  "tag": "",
}
`;

exports[`ol > li 2`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "content": "3. hello",
          "payload": {
            "listIndex": 3,
            "tag": "li",
          },
        },
        {
          "children": [],
          "content": "4. world",
          "payload": {
            "listIndex": 4,
            "tag": "li",
          },
        },
      ],
      "content": "",
    },
  ],
  "content": "",
}
`;

exports[`parseHtml 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "childrenLevel": 0,
                  "html": "<p>item 1</p>",
                  "id": 7,
                  "level": 9,
                  "parent": 6,
                  "tag": "li",
                },
                {
                  "children": [
                    {
                      "children": [
                        {
                          "children": [],
                          "childrenLevel": 0,
                          "html": "<p>item 3</p>",
                          "id": 10,
                          "level": 9,
                          "parent": 9,
                          "tag": "li",
                        },
                        {
                          "children": [],
                          "childrenLevel": 0,
                          "html": "<p>item 4</p>",
                          "id": 11,
                          "level": 9,
                          "parent": 9,
                          "tag": "li",
                        },
                      ],
                      "childrenLevel": 9,
                      "html": "",
                      "id": 9,
                      "level": 8,
                      "parent": 8,
                      "tag": "ul",
                    },
                  ],
                  "childrenLevel": 8,
                  "html": "<p>item 2</p>",
                  "id": 8,
                  "level": 9,
                  "parent": 6,
                  "tag": "li",
                },
              ],
              "childrenLevel": 9,
              "html": "",
              "id": 6,
              "level": 8,
              "parent": 5,
              "tag": "ul",
            },
            {
              "children": [
                {
                  "children": [],
                  "childrenLevel": 0,
                  "data": {
                    "listIndex": 1,
                  },
                  "html": "1. item 5",
                  "id": 13,
                  "level": 9,
                  "parent": 12,
                  "tag": "li",
                },
                {
                  "children": [],
                  "childrenLevel": 0,
                  "data": {
                    "listIndex": 2,
                  },
                  "html": "2. item 6",
                  "id": 14,
                  "level": 9,
                  "parent": 12,
                  "tag": "li",
                },
                {
                  "children": [
                    {
                      "children": [
                        {
                          "children": [
                            {
                              "children": [
                                {
                                  "children": [],
                                  "childrenLevel": 0,
                                  "html": "item 7.1.1",
                                  "id": 19,
                                  "level": 9,
                                  "parent": 18,
                                  "tag": "li",
                                },
                                {
                                  "children": [],
                                  "childrenLevel": 0,
                                  "html": "item 7.1.2",
                                  "id": 20,
                                  "level": 9,
                                  "parent": 18,
                                  "tag": "li",
                                },
                              ],
                              "childrenLevel": 9,
                              "html": "",
                              "id": 18,
                              "level": 8,
                              "parent": 17,
                              "tag": "ul",
                            },
                          ],
                          "childrenLevel": 8,
                          "html": "item 7.1",
                          "id": 17,
                          "level": 9,
                          "parent": 16,
                          "tag": "li",
                        },
                        {
                          "children": [],
                          "childrenLevel": 0,
                          "html": "item 7.2",
                          "id": 21,
                          "level": 9,
                          "parent": 16,
                          "tag": "li",
                        },
                      ],
                      "childrenLevel": 9,
                      "html": "",
                      "id": 16,
                      "level": 8,
                      "parent": 15,
                      "tag": "ul",
                    },
                  ],
                  "childrenLevel": 8,
                  "data": {
                    "listIndex": 3,
                  },
                  "html": "3. item 7",
                  "id": 15,
                  "level": 9,
                  "parent": 12,
                  "tag": "li",
                },
              ],
              "childrenLevel": 9,
              "html": "",
              "id": 12,
              "level": 8,
              "parent": 5,
              "tag": "ol",
            },
          ],
          "childrenLevel": 8,
          "comments": [
            "foldAll",
          ],
          "data": {
            "id": "h2",
          },
          "html": "heading 2",
          "id": 5,
          "level": 2,
          "parent": 1,
          "tag": "h2",
        },
        {
          "children": [
            {
              "children": undefined,
              "childrenLevel": 0,
              "html": "<table>
<tbody><tr>
<td>
<ul><li>this list is ignored</li></ul>
</td>
</tr>
</tbody></table>",
              "id": 23,
              "level": 7,
              "parent": 22,
              "tag": "table",
            },
            {
              "children": undefined,
              "childrenLevel": 0,
              "html": "<img src="image1.png">",
              "id": 24,
              "level": 7,
              "parent": 22,
              "tag": "img",
            },
            {
              "children": undefined,
              "childrenLevel": 0,
              "html": "<img src="image2.png">",
              "id": 25,
              "level": 7,
              "parent": 22,
              "tag": "img",
            },
          ],
          "childrenLevel": 7,
          "html": "A table",
          "id": 22,
          "level": 2,
          "parent": 1,
          "tag": "h2",
        },
      ],
      "childrenLevel": 2,
      "data": {
        "id": "h1",
      },
      "html": "heading 1",
      "id": 1,
      "level": 1,
      "parent": 0,
      "tag": "h1",
    },
  ],
  "childrenLevel": 1,
  "html": "",
  "id": 0,
  "level": 0,
  "parent": 0,
  "tag": "",
}
`;

exports[`parseHtml 2`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "content": "<p>item 1</p>",
                },
                {
                  "children": [
                    {
                      "children": [
                        {
                          "children": [],
                          "content": "<p>item 3</p>",
                        },
                        {
                          "children": [],
                          "content": "<p>item 4</p>",
                        },
                      ],
                      "content": "",
                    },
                  ],
                  "content": "<p>item 2</p>",
                },
              ],
              "content": "",
            },
            {
              "children": [
                {
                  "children": [],
                  "content": "1. item 5",
                  "payload": {
                    "listIndex": 1,
                    "tag": "li",
                  },
                },
                {
                  "children": [],
                  "content": "2. item 6",
                  "payload": {
                    "listIndex": 2,
                    "tag": "li",
                  },
                },
                {
                  "children": [
                    {
                      "children": [
                        {
                          "children": [
                            {
                              "children": [
                                {
                                  "children": [],
                                  "content": "item 7.1.1",
                                },
                                {
                                  "children": [],
                                  "content": "item 7.1.2",
                                },
                              ],
                              "content": "",
                            },
                          ],
                          "content": "item 7.1",
                        },
                        {
                          "children": [],
                          "content": "item 7.2",
                        },
                      ],
                      "content": "",
                    },
                  ],
                  "content": "3. item 7",
                  "payload": {
                    "listIndex": 3,
                    "tag": "li",
                  },
                },
              ],
              "content": "",
            },
          ],
          "content": "heading 2",
          "payload": {
            "fold": 2,
            "id": "h2",
            "tag": "h2",
          },
        },
        {
          "children": [
            {
              "children": [],
              "content": "<table>
<tbody><tr>
<td>
<ul><li>this list is ignored</li></ul>
</td>
</tr>
</tbody></table>",
            },
            {
              "children": [],
              "content": "<img src="image1.png">",
            },
            {
              "children": [],
              "content": "<img src="image2.png">",
            },
          ],
          "content": "A table",
        },
      ],
      "content": "heading 1",
      "payload": {
        "id": "h1",
        "tag": "h1",
      },
    },
  ],
  "content": "",
}
`;

exports[`parseHtml with data 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "childrenLevel": 0,
                  "data": {
                    "lines": "1,2",
                  },
                  "html": "l1.1",
                  "id": 4,
                  "level": 9,
                  "parent": 3,
                  "tag": "li",
                },
                {
                  "children": [
                    {
                      "children": [
                        {
                          "children": [],
                          "childrenLevel": 0,
                          "data": {
                            "lines": "3,4",
                          },
                          "html": "l1.2.1",
                          "id": 7,
                          "level": 9,
                          "parent": 6,
                          "tag": "li",
                        },
                      ],
                      "childrenLevel": 9,
                      "data": {
                        "lines": "3,4",
                      },
                      "html": "",
                      "id": 6,
                      "level": 8,
                      "parent": 5,
                      "tag": "ul",
                    },
                  ],
                  "childrenLevel": 8,
                  "data": {
                    "lines": "2,4",
                  },
                  "html": "l1.2",
                  "id": 5,
                  "level": 9,
                  "parent": 3,
                  "tag": "li",
                },
              ],
              "childrenLevel": 9,
              "data": {
                "lines": "1,4",
              },
              "html": "",
              "id": 3,
              "level": 8,
              "parent": 2,
              "tag": "ul",
            },
          ],
          "childrenLevel": 8,
          "data": {
            "lines": "0,4",
          },
          "html": "l1",
          "id": 2,
          "level": 9,
          "parent": 1,
          "tag": "li",
        },
      ],
      "childrenLevel": 9,
      "data": {
        "lines": "0,4",
      },
      "html": "",
      "id": 1,
      "level": 8,
      "parent": 0,
      "tag": "ul",
    },
  ],
  "childrenLevel": 8,
  "html": "",
  "id": 0,
  "level": 0,
  "parent": 0,
  "tag": "",
}
`;

exports[`parseHtml with data 2`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [
            {
              "children": [
                {
                  "children": [],
                  "content": "l1.1",
                  "payload": {
                    "lines": "1,2",
                    "tag": "li",
                  },
                },
                {
                  "children": [
                    {
                      "children": [
                        {
                          "children": [],
                          "content": "l1.2.1",
                          "payload": {
                            "lines": "3,4",
                            "tag": "li",
                          },
                        },
                      ],
                      "content": "",
                      "payload": {
                        "lines": "3,4",
                        "tag": "ul",
                      },
                    },
                  ],
                  "content": "l1.2",
                  "payload": {
                    "lines": "2,4",
                    "tag": "li",
                  },
                },
              ],
              "content": "",
              "payload": {
                "lines": "1,4",
                "tag": "ul",
              },
            },
          ],
          "content": "l1",
          "payload": {
            "lines": "0,4",
            "tag": "li",
          },
        },
      ],
      "content": "",
      "payload": {
        "lines": "0,4",
        "tag": "ul",
      },
    },
  ],
  "content": "",
}
`;
